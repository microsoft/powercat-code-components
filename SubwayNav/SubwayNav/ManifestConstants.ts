export const enum ManifestPropertyNames {
    dataset = 'dataset',
}

export const enum ItemColumns {
    Label = 'ItemLabel',
    Key = 'ItemKey',
    State = 'ItemState',
    Disabled = 'ItemDisabled',
    VisuallyDisabled = 'ItemVisuallyDisabled',
    ParentKey = 'ParentItemKey',
}

export const enum InputEvents {
    SetFocus = 'SetFocus',
}

export const enum InputProperties {
    InputEvent = 'InputEvent',
    SelectedKey = 'SelectedKey',
}
