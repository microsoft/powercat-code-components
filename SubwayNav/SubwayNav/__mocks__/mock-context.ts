/* istanbul ignore file */

export class MockContext<T> implements ComponentFramework.Context<T> {
    constructor(parameters: T) {
        this.parameters = parameters;
        this.mode = {
            allocatedHeight: -1,
            allocatedWidth: -1,
            isControlDisabled: false,
            isVisible: true,
            label: '',
            setControlState: jest.fn(),
            setFullScreen: jest.fn(),
            trackContainerResize: jest.fn(),
        };
        this.client = {
            disableScroll: false,
            getClient: jest.fn(),
            getFormFactor: jest.fn(),
            isOffline: jest.fn(),
        };

        // Canvas apps currently assigns a positive tab-index
        // so we must use this property to assign a positive tab-index also
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        (this as any).accessibility = { assignedTabIndex: 0 };
    }
    client: ComponentFramework.Client;
    device: ComponentFramework.Device;
    factory: ComponentFramework.Factory;
    formatting: ComponentFramework.Formatting;
    mode: ComponentFramework.Mode;
    navigation: ComponentFramework.Navigation;
    resources: ComponentFramework.Resources;
    userSettings: ComponentFramework.UserSettings;
    utils: ComponentFramework.Utility;
    webAPI: ComponentFramework.WebApi;
    parameters: T;
    updatedProperties: string[] = [];
}

export class MockState implements ComponentFramework.Dictionary {}

export class MockStringProperty implements ComponentFramework.PropertyTypes.StringProperty {
    constructor(raw?: string | null, formatted?: string | undefined) {
        this.raw = raw ?? null;
        this.formatted = formatted;
    }
    raw: string | null;
    attributes?: ComponentFramework.PropertyHelper.FieldPropertyMetadata.StringMetadata | undefined;
    error: boolean;
    errorMessage: string;
    formatted?: string | undefined;
    security?: ComponentFramework.PropertyHelper.SecurityValues | undefined;
    type: string;
}

export class MockWholeNumberProperty implements ComponentFramework.PropertyTypes.WholeNumberProperty {
    constructor(raw?: number | null, formatted?: string | undefined) {
        this.raw = raw ?? null;
        this.formatted = formatted;
    }
    attributes?: ComponentFramework.PropertyHelper.FieldPropertyMetadata.WholeNumberMetadata | undefined;
    raw: number | null;
    error: boolean;
    errorMessage: string;
    formatted?: string | undefined;
    security?: ComponentFramework.PropertyHelper.SecurityValues | undefined;
    type: string;
}

export class MockEnumProperty<T> implements ComponentFramework.PropertyTypes.EnumProperty<T> {
    constructor(raw?: T, type?: string) {
        if (raw) this.raw = raw;
        if (type) this.type = type;
    }
    type: string;
    raw: T;
}

export class MockTwoOptionsProperty implements ComponentFramework.PropertyTypes.TwoOptionsProperty {
    constructor(raw?: boolean) {
        if (raw) this.raw = raw;
    }
    raw: boolean;
    attributes?: ComponentFramework.PropertyHelper.FieldPropertyMetadata.TwoOptionMetadata | undefined;
    error: boolean;
    errorMessage: string;
    formatted?: string | undefined;
    security?: ComponentFramework.PropertyHelper.SecurityValues | undefined;
    type: string;
}
