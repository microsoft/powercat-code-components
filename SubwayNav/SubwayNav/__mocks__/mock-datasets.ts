/* istanbul ignore file */

export class MockDataSet implements ComponentFramework.PropertyTypes.DataSet {
    private rows: MockEntityRecord[] = [];
    constructor(rows: MockEntityRecord[]) {
        this.rows = rows;
        this.records = {};
        rows.forEach((r) => (this.records[r.id] = r));
        this.sortedRecordIds = rows.map((r) => r.id);
        this.paging = {
            setPageSize: jest.fn(),
            totalResultCount: 0,
            firstPageNumber: 0,
            hasNextPage: false,
            hasPreviousPage: false,
            lastPageNumber: 0,
            loadExactPage: jest.fn(),
            loadNextPage: jest.fn(),
            loadPreviousPage: jest.fn(),
            pageSize: 0,
            reset: jest.fn(),
        };
    }
    addColumn = jest.fn();
    columns: ComponentFramework.PropertyHelper.DataSetApi.Column[] = [];
    error: boolean;
    errorMessage: string;
    filtering: ComponentFramework.PropertyHelper.DataSetApi.Filtering;
    linking: ComponentFramework.PropertyHelper.DataSetApi.Linking;
    loading: boolean;
    paging: ComponentFramework.PropertyHelper.DataSetApi.Paging;
    records: { [id: string]: ComponentFramework.PropertyHelper.DataSetApi.EntityRecord };
    sortedRecordIds: string[];
    sorting: ComponentFramework.PropertyHelper.DataSetApi.SortStatus[];
    clearSelectedRecordIds = jest.fn();
    getSelectedRecordIds = jest.fn();
    getTargetEntityType = jest.fn();
    getTitle = jest.fn();
    getViewId = jest.fn();
    openDatasetItem = jest.fn();
    refresh = jest.fn();
    setSelectedRecordIds = jest.fn();
}

export class MockColumn implements ComponentFramework.PropertyHelper.DataSetApi.Column {
    name: string;
    displayName: string;
    dataType!: string;
    alias!: string;
    order!: number;
    visualSizeFactor!: number;
    isHidden?: boolean | undefined;
    isPrimary?: boolean | undefined;
    disableSorting?: boolean | undefined;
    constructor(name: string, displayName: string) {
        this.name = name;
        this.displayName = displayName;
    }
}

type valueType =
    | string
    | number
    | boolean
    | Date
    | number[]
    | ComponentFramework.EntityReference
    | ComponentFramework.EntityReference[]
    | ComponentFramework.LookupValue
    | ComponentFramework.LookupValue[];

export class MockEntityRecord implements ComponentFramework.PropertyHelper.DataSetApi.EntityRecord {
    values: Record<string, valueType>;
    id: string;
    constructor(id: string, values: Record<string, valueType>) {
        this.values = values;
        this.id = id;
    }
    getFormattedValue(columnName: string): string {
        return this.values[columnName] as string;
    }
    getRecordId(): string {
        return this.id;
    }
    getValue(columnName: string): valueType {
        return this.values[columnName];
    }
    getNamedReference = jest.fn();
}

export function getData(records: ComponentFramework.PropertyHelper.DataSetApi.EntityRecord[]): {
    sortedRecordIds: string[];
    records: Record<string, ComponentFramework.PropertyHelper.DataSetApi.EntityRecord>;
} {
    const sortedRecordIds: string[] = [];
    const recordsOut: Record<string, ComponentFramework.PropertyHelper.DataSetApi.EntityRecord> = {};

    for (const r of records) {
        sortedRecordIds.push(r.getRecordId());
        recordsOut[r.getRecordId()] = r;
    }
    return {
        sortedRecordIds: sortedRecordIds,
        records: recordsOut,
    };
}
