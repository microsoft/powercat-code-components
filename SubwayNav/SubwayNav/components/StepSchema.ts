export interface IOutputSchemaMap {
    [propertyKey: string]: Record<string, unknown>;
}

export class StepSchema {
    public static getStepSchema(): IOutputSchemaMap {
        return stepSchema;
    }
}

const stepDetails = {
    type: 'object',
    properties: {
        ItemKey: { type: 'string' },
        ItemLabel: { type: 'string' },
        ItemState: { type: 'string' },
        ItemDisabled: { type: 'boolean' },
        ItemVisuallyDisabled: { type: 'boolean' },
        ParentItemKey: { type: 'string' },
    },
};

const stepSchema = {
    Steps: {
        type: 'array',
        items: stepDetails,
    },
};
