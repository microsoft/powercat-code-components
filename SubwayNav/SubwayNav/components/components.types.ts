import { ISubwayNavNodeProps } from '../utilities/subway-nav/subway-node.types';

export interface ISubNavItem {
    id: string;
    parentId?: string;
    index?: number;
    label: string;
    state: string;
    key: string;
    clickable?: boolean;
    disabled?: boolean;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    data: any;
    visuallyDisabled?: boolean;
}

export interface ICustomSubwayNavProps {
    ItemKey: string;
    ItemLabel: string;
    ItemState?: string;
    ItemDisabled?: boolean;
    ParentItemKey?: string;
    ItemVisuallyDisabled?: boolean;
}

export interface ISubNavProps {
    width?: number;
    height?: number;
    items: ISubNavItem[];
    onSelected: (item: ISubNavItem, steps: ISubwayNavNodeProps[]) => void;
    themeJSON?: string;
    ariaLabel?: string;
    setFocus?: string;
    tabIndex?: number;
    maxDisplayedItems?: number;
    overflowIndex?: number;
    applyDarkTheme?: boolean;
    disabled: boolean;
    wizardComplete: string;
}
