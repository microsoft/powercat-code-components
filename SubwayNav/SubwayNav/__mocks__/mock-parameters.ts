/* istanbul ignore file */

import { IInputs } from '../generated/ManifestTypes';
import { MockEnumProperty, MockStringProperty, MockTwoOptionsProperty } from './mock-context';
import { MockDataSet } from './mock-datasets';

export function getMockParameters(): IInputs {
    return {
        AccessibilityLabel: new MockStringProperty(),
        InputEvent: new MockStringProperty(),
        Theme: new MockStringProperty(),
        items: new MockDataSet([]),
        ApplyDarkTheme: new MockTwoOptionsProperty(),
        WizardCompleteorError: new MockEnumProperty(),
        StepsSchema: new MockStringProperty(),
    };
}
