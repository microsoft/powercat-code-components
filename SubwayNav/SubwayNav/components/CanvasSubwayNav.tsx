import * as React from 'react';
import { useState } from 'react';
import { ThemeProvider, createTheme, ITheme, IFocusZoneProps } from '@fluentui/react';
import { ISubwayNavNodeProps, SubwayNavNodeState } from '../utilities/subway-nav/subway-node.types';
import { goToStepById, completeAllSteps, errorAllSteps } from '../utilities/utilities';
import {
    SubwayNav as CustomSubwayNav,
    SubwayNavNoAnimation as CustomSubwayNavNoAnimation,
} from '../utilities/subway-nav/subway-nav';
import { ISubNavItem, ISubNavProps } from './components.types';

import { M365Styles, IM365ExtendedSemanticColors } from '../utilities/customizations/src';
import { useAsync, usePrevious } from '@fluentui/react-hooks';
import { getSubwayNavNodeState } from './DatasetMapping';
import { PPACActualLightTheme, PPACActualDarkTheme } from '../utilities/themes';
import { ISubwayNavProps } from '../utilities/subway-nav/subway-nav.types';

// reference : https://admincontrolsdemoapps.blob.core.windows.net/demo-app/latest/DemoApp/index.html#/examples/subwaynav

const ariaLabelStrings = {
    NotStarted: 'Not Started',
    Current: 'Current',
    CurrentWithSubSteps: 'Current With Sub Steps',
    Completed: 'Completed',
    ViewedNotCompleted: 'Viewed Not Completed',
    Unsaved: 'Unsaved',
    Skipped: 'Skipped',
    Error: 'Error',
    WizardComplete: 'Wizard Complete',
    Custom: 'Custom',
};

export const CanvasSubwayNav = React.memo((props: ISubNavProps): React.ReactElement => {
    const {
        items,
        themeJSON,
        onSelected,
        setFocus,
        applyDarkTheme,
        tabIndex,
        disabled,
        wizardComplete,
        showAnimation,
    } = props;
    const [isNavCompleteOrError, setIsNavCompleteOrError] = useState(false);
    let allSteps: ISubwayNavNodeProps[];
    const prevItems = usePrevious(items);
    const prevNavState = usePrevious(wizardComplete);

    const theme = React.useMemo(() => {
        try {
            // Using PPAC theme of Admin Controls
            const copyofM365Theme: ITheme = applyDarkTheme ? PPACActualDarkTheme : PPACActualLightTheme;
            const semanticColorsCopy: Partial<IM365ExtendedSemanticColors> = copyofM365Theme.semanticColors;

            // Overiding semanticcolors specified for Subwaynav Steps using custom themeJSON
            if (themeJSON) {
                const themePrimary = JSON.parse(themeJSON).palette.themePrimary;
                semanticColorsCopy.stepNotStarted = JSON.parse(themeJSON).palette.neutralTertiaryAlt;
                semanticColorsCopy.stepPressed = themePrimary;
                semanticColorsCopy.stepHover = JSON.parse(themeJSON).palette.themeDark;
                semanticColorsCopy.stepCompleted = themePrimary;
                semanticColorsCopy.stepCurrent = themePrimary;
                semanticColorsCopy.stepModifierBorder = themePrimary;
            }
            return themeJSON
                ? createTheme({
                      palette: { ...JSON.parse(themeJSON).palette },
                      semanticColors: semanticColorsCopy,
                      components: M365Styles,
                  })
                : copyofM365Theme;
        } catch (ex) {
            /* istanbul ignore next */
            console.error('Cannot parse theme', ex);
            // so that, control doesn't break
            return applyDarkTheme ? PPACActualDarkTheme : PPACActualLightTheme;
        }
    }, [themeJSON, applyDarkTheme]);

    const componentRef = React.useRef<ISubwayNavNodeProps>(null);
    const async = useAsync();
    React.useEffect(() => {
        if (setFocus && setFocus !== '' && componentRef) {
            async.requestAnimationFrame(() => {
                (componentRef as React.RefObject<ISubwayNavNodeProps>).current?.focus();
            });
        }
    }, [setFocus, componentRef, async]);

    const focusZoneProps = React.useMemo(() => {
        return {
            tabIndex: tabIndex,
            shouldFocusInnerElementWhenReceivedFocus: true,
        } as IFocusZoneProps;
    }, [tabIndex]);

    // eslint-disable-next-line react-hooks/exhaustive-deps
    function getSteps(items: ISubNavItem[], onClickStep?: (props: ISubwayNavNodeProps) => void): ISubwayNavNodeProps[] {
        allSteps = items
            .filter((item) => !item.parentId)
            .map((group) => {
                const subSteps = items
                    .filter((subItem) => subItem.parentId === group.key)
                    .map((item: ISubNavItem) => {
                        return {
                            id: item.id,
                            label: item.label,
                            state:
                                wizardComplete !== 'None'
                                    ? getSubwayNavNodeState(wizardComplete)
                                    : getSubwayNavNodeState(item.state),
                            disabled: item.disabled ?? false,
                            parentId: item.parentId,
                            onClickStep,
                            index: 10,
                            isVisuallyDisabled: item.visuallyDisabled ?? false,
                            itemIcon: item.itemIcon,
                            itemColor: item.itemColor,
                        };
                    });
                return {
                    id: group.id,
                    key: group.key,
                    label: group.label,
                    state:
                        wizardComplete !== 'None'
                            ? getSubwayNavNodeState(wizardComplete)
                            : getSubwayNavNodeState(group.state),
                    data: group,
                    ...(subSteps.length > 0 && { subSteps: subSteps }),
                    disabled: group.disabled ?? false,
                    onClickStep,
                    index: 10,
                    isVisuallyDisabled: group.visuallyDisabled ?? false,
                    itemIcon: group.itemIcon,
                    itemColor: group.itemColor,
                };
            }) as unknown as ISubwayNavNodeProps[];
        return allSteps;
    }

    // eslint-disable-next-line react-hooks/exhaustive-deps
    const handleClickStep = (step: ISubwayNavNodeProps): void => {
        if (step) {
            const selectedStep = items.find((item: ISubNavItem) => item.id === step?.id);
            const nextSteps = goToStepById(allSteps, step.id, SubwayNavNodeState.Unsaved).steps;
            setSteps(nextSteps);
            if (selectedStep) onSelected(selectedStep, nextSteps);
        }
    };

    const [steps, setSteps] = useState<ISubwayNavNodeProps[]>(getSteps(items, handleClickStep));

    React.useEffect(() => {
        //If item or Wizard state change
        if (prevItems !== items || prevNavState !== wizardComplete) {
            setSteps(getSteps(items, handleClickStep));
        }
        //Set Wizard State
        if (wizardComplete !== 'None') {
            if (wizardComplete === 'Complete') {
                setIsNavCompleteOrError(true);
                completeAllSteps(steps);
            } else {
                setIsNavCompleteOrError(false);
                errorAllSteps(steps);
            }
        }
    }, [wizardComplete, steps, items, getSteps, handleClickStep, prevItems, prevNavState]);

    const subwaynavProps = {
        disabled: disabled,
        focusZoneProps: focusZoneProps,
        stateAriaLabels: ariaLabelStrings,
        steps: steps,
        // This is required to override width styles in custom pages
        styles: { root: { width: props.width } },
        ...(wizardComplete !== 'None' && { wizardComplete: isNavCompleteOrError }),
    } as ISubwayNavProps;

    return (
        <ThemeProvider theme={theme}>
            {showAnimation ? (
                <CustomSubwayNav {...subwaynavProps} />
            ) : (
                <CustomSubwayNavNoAnimation {...subwaynavProps} />
            )}
        </ThemeProvider>
    );
});
CanvasSubwayNav.displayName = 'CanvasSubwayNav';
