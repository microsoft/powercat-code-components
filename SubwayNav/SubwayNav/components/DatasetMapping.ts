import { ItemColumns } from '../ManifestConstants';
import { ISubNavItem, ICustomSubwayNavProps } from './components.types';
import { SubwayNavNodeState, ISubwayNavNodeProps } from '../utilities/subway-nav/subway-node.types';

export function getItemsFromDataset(dataset: ComponentFramework.PropertyTypes.DataSet): ISubNavItem[] {
    if (dataset.error || dataset.paging.totalResultCount === undefined) {
        // Dataset is not defined so return dummy items
        return [getDummyAction('1'), getDummyAction('2'), getDummyAction('3'), getDummyAction('4')];
    }
    const keyIndex: Record<string, number> = {};
    return dataset.sortedRecordIds.map((id) => {
        const record = dataset.records[id];
        // Prevent duplicate keys by appending the duplicate index
        let key = record.getValue(ItemColumns.Key) as string;
        if (keyIndex[key] !== undefined) {
            keyIndex[key]++;
            key += `_${keyIndex[key]}`;
        } else keyIndex[key] = 1;
        return {
            id: record.getRecordId(),
            key: record.getValue(ItemColumns.Key),
            label: record.getFormattedValue(ItemColumns.Label),
            state: record.getValue(ItemColumns.State),
            disabled: record.getValue(ItemColumns.Disabled),
            parentId: record.getValue(ItemColumns.ParentKey),
            visuallyDisabled: record.getValue(ItemColumns.VisuallyDisabled),
            data: record,
        } as ISubNavItem;
    });
}

export function getDatasetfromItems(steps: ISubwayNavNodeProps[]): ICustomSubwayNavProps[] {
    let stepDataSet: ICustomSubwayNavProps[] = [];
    const parentstepDateSet = steps.map((step: ISubwayNavNodeProps) => {
        if (Array.isArray(step.subSteps) && step.subSteps.length) {
            stepDataSet = stepDataSet.concat(
                step.subSteps.map((subStep) => {
                    return {
                        ItemKey: subStep.id,
                        ItemLabel: subStep.label,
                        ItemState: subStep.state,
                        ItemDisabled: subStep.disabled,
                        ParentItemKey: subStep.parentId,
                        ItemVisuallyDisabled: subStep.isVisuallyDisabled,
                    } as ICustomSubwayNavProps;
                }),
            );
        }
        return {
            ItemKey: step.id,
            ItemLabel: step.label,
            ItemState: step.state,
            ItemDisabled: step.disabled,
            ParentItemKey: step.parentId,
            ItemVisuallyDisabled: step.isVisuallyDisabled,
        } as ICustomSubwayNavProps;
    });
    return [...parentstepDateSet, ...stepDataSet];
}

function getDummyAction(id: string): ISubNavItem {
    return {
        id: id,
        label: 'Node ' + id,
        key: id,
        parentId: id === '5' ? '2' : undefined,
        state: id === '1' ? 'Current' : 'Not Started',
    } as ISubNavItem;
}

export function getSubwayNavNodeState(state: string): SubwayNavNodeState {
    switch (state) {
        case 'NotStarted':
            return SubwayNavNodeState.NotStarted;
        case 'Current':
            return SubwayNavNodeState.Current;
        case 'Completed':
            return SubwayNavNodeState.Completed;
        case 'ViewedNotCompleted':
            return SubwayNavNodeState.ViewedNotCompleted;
        case 'Unsaved':
            return SubwayNavNodeState.Unsaved;
        case 'Skipped':
            return SubwayNavNodeState.Skipped;
        case 'Error':
            return SubwayNavNodeState.Error;
        case 'CurrentWithSubSteps':
            return SubwayNavNodeState.CurrentWithSubSteps;
        case 'WizardComplete':
            return SubwayNavNodeState.WizardComplete;
        default:
            return SubwayNavNodeState.NotStarted;
    }
}
