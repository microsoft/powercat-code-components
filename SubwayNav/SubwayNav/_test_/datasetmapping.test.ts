import { getItemsFromDataset } from '../components/DatasetMapping';
import { ItemColumns } from '../ManifestConstants';
import { MockDataSet, MockEntityRecord } from '../__mocks__/mock-datasets';

describe('DatasetMapping', () => {
    beforeEach(() => jest.clearAllMocks());

    it('returns correct props', () => {
        const items = [
            new MockEntityRecord('1', {
                [ItemColumns.Key]: 'Item 1',
                [ItemColumns.Label]: 'item1',
                [ItemColumns.State]: 'Current',
            }),
            new MockEntityRecord('2', {
                [ItemColumns.Key]: 'Item 2',
                [ItemColumns.Label]: 'item2',
                [ItemColumns.State]: 'Not Started',
            }),
            new MockEntityRecord('3', {
                [ItemColumns.Key]: 'Item 3',
                [ItemColumns.Label]: 'item3',
                [ItemColumns.State]: 'Completed',
            }),
            new MockEntityRecord('4', {
                [ItemColumns.Key]: 'Item 4',
                [ItemColumns.Label]: 'item4',
                [ItemColumns.State]: 'ViewedNotCompleted',
            }),
            new MockEntityRecord('5', {
                [ItemColumns.Key]: 'Item 4',
                [ItemColumns.Label]: 'item4',
                [ItemColumns.State]: 'Error',
            }),
            new MockEntityRecord('6', {
                [ItemColumns.Key]: 'Item 4',
                [ItemColumns.Label]: 'item4',
                [ItemColumns.State]: 'Skipped',
            }),
            new MockEntityRecord('7', {
                [ItemColumns.Key]: 'Item 4',
                [ItemColumns.Label]: 'item4',
                [ItemColumns.State]: 'Unsaved',
            }),
        ];

        const actions = getItemsFromDataset(new MockDataSet(items));
        expect(actions).toMatchSnapshot();
    });
});
