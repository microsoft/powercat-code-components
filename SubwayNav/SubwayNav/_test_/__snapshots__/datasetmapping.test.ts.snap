// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatasetMapping returns correct props 1`] = `
Array [
  Object {
    "data": MockEntityRecord {
      "getNamedReference": [MockFunction],
      "id": "1",
      "values": Object {
        "ItemKey": "Item 1",
        "ItemLabel": "item1",
        "ItemState": "Current",
      },
    },
    "disabled": undefined,
    "id": "1",
    "key": "Item 1",
    "label": "item1",
    "parentId": undefined,
    "state": "Current",
    "visuallyDisabled": undefined,
  },
  Object {
    "data": MockEntityRecord {
      "getNamedReference": [MockFunction],
      "id": "2",
      "values": Object {
        "ItemKey": "Item 2",
        "ItemLabel": "item2",
        "ItemState": "Not Started",
      },
    },
    "disabled": undefined,
    "id": "2",
    "key": "Item 2",
    "label": "item2",
    "parentId": undefined,
    "state": "Not Started",
    "visuallyDisabled": undefined,
  },
  Object {
    "data": MockEntityRecord {
      "getNamedReference": [MockFunction],
      "id": "3",
      "values": Object {
        "ItemKey": "Item 3",
        "ItemLabel": "item3",
        "ItemState": "Completed",
      },
    },
    "disabled": undefined,
    "id": "3",
    "key": "Item 3",
    "label": "item3",
    "parentId": undefined,
    "state": "Completed",
    "visuallyDisabled": undefined,
  },
  Object {
    "data": MockEntityRecord {
      "getNamedReference": [MockFunction],
      "id": "4",
      "values": Object {
        "ItemKey": "Item 4",
        "ItemLabel": "item4",
        "ItemState": "ViewedNotCompleted",
      },
    },
    "disabled": undefined,
    "id": "4",
    "key": "Item 4",
    "label": "item4",
    "parentId": undefined,
    "state": "ViewedNotCompleted",
    "visuallyDisabled": undefined,
  },
  Object {
    "data": MockEntityRecord {
      "getNamedReference": [MockFunction],
      "id": "5",
      "values": Object {
        "ItemKey": "Item 4",
        "ItemLabel": "item4",
        "ItemState": "Error",
      },
    },
    "disabled": undefined,
    "id": "5",
    "key": "Item 4",
    "label": "item4",
    "parentId": undefined,
    "state": "Error",
    "visuallyDisabled": undefined,
  },
  Object {
    "data": MockEntityRecord {
      "getNamedReference": [MockFunction],
      "id": "6",
      "values": Object {
        "ItemKey": "Item 4",
        "ItemLabel": "item4",
        "ItemState": "Skipped",
      },
    },
    "disabled": undefined,
    "id": "6",
    "key": "Item 4",
    "label": "item4",
    "parentId": undefined,
    "state": "Skipped",
    "visuallyDisabled": undefined,
  },
  Object {
    "data": MockEntityRecord {
      "getNamedReference": [MockFunction],
      "id": "7",
      "values": Object {
        "ItemKey": "Item 4",
        "ItemLabel": "item4",
        "ItemState": "Unsaved",
      },
    },
    "disabled": undefined,
    "id": "7",
    "key": "Item 4",
    "label": "item4",
    "parentId": undefined,
    "state": "Unsaved",
    "visuallyDisabled": undefined,
  },
]
`;
